### PR TITLE
Clipping support

### PR DESCRIPTION
**Pull request by Jhonnyg**
This PR adds clipping support to the extension:

image

### CODE REVIEW

File: defold-spine/src/gui_node_spine.cpp [hunk 1]
```diff
@@ -36,7 +37,7 @@ static const dmhash_t SPINE_SCENE_SUFFIX        = dmHashString64(".spinescenec")
 
 struct GuiNodeTypeContext
 {
-    // In case we need something later. Here for visibility
+    spSkeletonClipping* m_SkeletonClipper;
```
Comment (Jhonnyg) [inline defold-spine/src/gui_node_spine.cpp:40]: This actually allocates a bit of memory, so we typically don't want to allocate on of these per node. We just need one of these temporarily when generating the vertices

File: defold-spine/commonsrc/vertices.cpp [hunk 2]
```diff
@@ -111,36 +126,109 @@ void GetSkeletonBounds(const spSkeleton* skeleton, SpineModelBounds& bounds)
...
+
+        if (num_tri_vertices > *out_max_triangle_count)
+        {
+            *out_max_triangle_count = num_tri_vertices;
+        }
+    }
 }
 
-uint32_t CalcVertexBufferSize(const spSkeleton* skeleton, uint32_t* out_max_triangle_count)
+static void CalcAndAddVertexBufferAttachmentByClipper(dmArray<float>& scratch, spSlot* slot, spAttachment* attachment, spSkeletonClipping* skeleton_clipper, uint32_t* out_count, uint32_t* out_max_triangle_count)
```
Comment (Jhonnyg) [inline defold-spine/commonsrc/vertices.cpp:150]: I wonder how slow this is (I don't have any good data to test with) - perhaps we should pass in the scratch buffer to the GenerateVertexData function and resize/reallocate as needed? We need to at least do this once for generating the vertex buffer size, but perhaps the second time could be avoided 🤔

File: defold-spine/commonsrc/vertices.cpp [hunk 3]
```diff
@@ -38,11 +42,20 @@ static uint32_t EnsureArrayFitsNumber(dmArray<T>& array, uint32_t num_to_add)
...
 
+template <typename T>
+static void EnsureArraySize(dmArray<T>& array, uint32_t size)
+{
+    if (array.Capacity() < size)
+    {
+        array.SetCapacity(size);
+    }
+    array.SetSize(size);
+}
```
Comment (Jhonnyg) [inline defold-spine/commonsrc/vertices.cpp:53]: I kinda feel this is a useful function to have in dmsdk, I see variants of this (and the one above) now and then.. Maybe some sort of array utility header in dmsdk?

File: defold-spine/commonsrc/vertices.cpp [hunk 4]
```diff
@@ -196,74 +281,84 @@ uint32_t GenerateVertexData(dmArray<SpineVertex>& vertex_buffer, const spSkeleto
...
+            uvs           = mesh->uvs;
+            indices       = mesh->triangles;
+            indices_count = mesh->trianglesCount;
+            color         = &mesh->color;
+            vertices      = scratch_vertex_floats.Begin();
+        }
+        else if (attachment->type == SP_ATTACHMENT_CLIPPING)
+        {
+            // Clipper setup is very similar to this:
+            // https://github.com/EsotericSoftware/spine-runtimes/blob/4.0/spine-sfml/c/src/spine/spine-sfml.cpp#L293
```
Comment (Jhonnyg) [inline defold-spine/commonsrc/vertices.cpp:332]: I was directed here from the main manual and have structured all the clipping calls according to how they do it, so I expect it to be correct :)

**Review by JCash — APPROVED** [review]
lgtm

**Review by AGulev — APPROVED** [review]

**Pull request merged**